src/types/tdigest.cc

```diff
@@ -0,0 +1,439 @@
+      if (index <= weight_sum) {
+        break;
+      }
+    }
+    DCHECK_LT(ci, td.size());
+
+    // deviation of index from the centroid center
+    double diff = index + td[ci].weight / 2 - weight_sum;
+
+    // index happen to be in a unit weight centroid
+    if (td[ci].weight == 1 && std::abs(diff) < 0.5) {
+      return td[ci].mean;
+    }
+
+    // find adjacent centroids for interpolation
+    uint32_t ci_left = ci, ci_right = ci;
+    if (diff > 0) {
+      if (ci_right == td.size() - 1) {
+        // index larger than center of last bin
+        DCHECK_EQ(weight_sum, total_weight_);
+        const Centroid* c = &td[ci_right];
+        DCHECK_GE(c->weight, 2);
+        return Lerp(c->mean, max_, diff / (c->weight / 2));
+      }
+      ++ci_right;
+    } else {
+      if (ci_left == 0) {
+        // index smaller than center of first bin
+        const Centroid* c = &td[0];
+        DCHECK_GE(c->weight, 2);
+        return Lerp(min_, c->mean, index / (c->weight / 2));
+      }
+      --ci_left;
+      diff += td[ci_left].weight / 2 + td[ci_right].weight / 2;
+    }
+
+    // interpolate from adjacent centroids
+    diff /= (td[ci_left].weight / 2 + td[ci_right].weight / 2);
+    return Lerp(td[ci_left].mean, td[ci_right].mean, diff);
+  }
+
+  double Mean() const {
+    double sum = 0;
+    for (const auto& centroid : tdigests_[current_]) {
+      sum += centroid.mean * centroid.weight;
+    }
+    return total_weight_ == 0 ? NAN : sum / total_weight_;
+  }
+
+  double TotalWeight() const { return total_weight_; }
+
+ private:
+  // must be declared before merger_, see constructor initialization list
+  const uint32_t delta_;
+
+  TDigestMerger<> merger_;
+  double total_weight_;
+  double min_, max_;
+
+  // ping-pong buffer holds two tdigests, size = 2 * delta * sizeof(Centroid)
+  std::vector<Centroid> tdigests_[2];
+  // index of active tdigest buffer, 0 or 1
+  int current_;
+};
+
+TDigest::TDigest(uint64_t delta) : impl_(std::make_unique<TDigestImpl>(delta)) { Reset({}); }
+
+void TDigest::Merge(const std::vector<TDigest>& others) {
+  if (others.empty()) {
+    return;
+  }
+
+  std::vector<const TDigestImpl*> impls;
+  impls.reserve(others.size());
+
+  std::transform(others.cbegin(), others.cend(), std::back_inserter(impls),
+                 [](const TDigest& i) { return i.impl_.get(); });
+
+  impl_->Merge(impls);
+}
+
+void TDigest::Reset(const CentroidsWithDelta& centroids_list) {
+  impl_->Reset(centroids_list.centroids, centroids_list.min, centroids_list.max, centroids_list.total_weight);
+}
+
+void TDigest::Reset() { impl_->Reset(); }
+
+CentroidsWithDelta TDigest::DumpCentroids() const {
+  auto centroids = impl_->Centroids();
+  return {
+      .centroids = std::move(centroids),
+      .delta = impl_->Delta(),
+      .min = impl_->Min(),
+      .max = impl_->Max(),
+      .total_weight = impl_->TotalWeight(),
+  };
+}
+
+void TDigest::Add(std::vector<double> items) { impl_->MergeInput(items); }
```

Comment:
Should the parameter be a reference?

Reply:
Hi @PragmaTwice ,

I have updated the public method as a const reference and the internal method tries to sort the vector, so I make it a pass by value. 😄

Best Regards,
Edward